File: poh5/poh5_4b.vb
module PaizaPOH5_4b
	dim SIZE as integer=9999999
	dim z(SIZE-1) as byte
	dim input_count as integer=0

	function myget() as integer
		dim r as integer
		while 48<=z(input_count) andalso z(input_count)<=57
			r=r*10+z(input_count)-48
			input_count+=1
		end while
		input_count+=1
		return r
	end function

	sub Main()
		Console.OpenStandardInput().Read(z,0,SIZE)
		dim W as integer=myget()
		dim H as integer=myget()
		dim Q as integer=myget()
		dim m(H,W) as integer
		dim imos(H+1,W+1) as integer
		for i as integer=0 to H-1
			for j as integer=0 to W-1
				m(i,j)=myget()
			next
		next
		dim s as integer
		for i as integer=0 to Q-1
			dim w1 as integer=myget()-1
			dim h1 as integer=myget()-1
			dim w2 as integer=myget()
			dim h2 as integer=myget()
			imos(h1,w1)+=1
			imos(h1,w2)-=1
			imos(h2,w1)-=1
			imos(h2,w2)+=1
		next
		for i as integer=0 to H-1
			for j as integer=1 to W-1
				imos(i,j)+=imos(i,j-1)
			next
		next
		for i as integer=1 to H-1
			for j as integer=0 to W-1
				imos(i,j)+=imos(i-1,j)
			next
		next
		for i as integer=0 to H-1
			for j as integer=0 to W-1
				if imos(i,j)>0
					s+=m(i,j)
				end if
			next
		next
		Console.WriteLine(s)
	end sub
end module

File: poh5/poh5_4b.cs
using System;
using System.Linq;
class PaizaPOH5_4b{
	const int SIZE=9999999;
	static byte[] z=new byte[SIZE];
	static int input_count=0;

	static int get(){
		int r=0;
		for(;'0'<=z[input_count]&&z[input_count]<='9';)r=r*10+z[input_count++]-'0';
		input_count++;
		return r;
	}
	static void Main(){
		Console.OpenStandardInput().Read(z,0,SIZE);
		int W=get(),H=get(),Q=get();
		int[,] m=new int[H,W];
		int[,] imos=new int[H+1,W+1];
		for(int h=0;h<H;h++){
			for(int w=0;w<W;w++){
				m[h,w]=get();
			}
		}
		int s=0;
		for(;Q-->0;){
			int w1=get()-1,h1=get()-1,w2=get(),h2=get();
			imos[h1,w1]+=1;
			imos[h1,w2]-=1;
			imos[h2,w1]-=1;
			imos[h2,w2]+=1;
		}
		for(int h=0;h<H;h++)for(int w=1;w<W;w++)imos[h,w]+=imos[h,w-1];
		for(int h=1;h<H;h++)for(int w=0;w<W;w++)imos[h,w]+=imos[h-1,w];
		for(int h=0;h<H;h++)for(int w=0;w<W;w++)if(imos[h,w]>0)s+=m[h,w];
		Console.WriteLine(s);
	}
}